File: app/src/main/java/org/lineageos/jelly/favorite/FavoriteAdapter.java
/*
 * Copyright (C) 2017 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.jelly.favorite;

import android.content.Context;
import android.database.Cursor;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import org.lineageos.jelly.R;

class FavoriteAdapter extends RecyclerView.Adapter<FavoriteHolder> {
    private final Context mContext;
    private Cursor mCursor;

    private int mIdColumnIndex;
    private int mTitleColumnIndex;
    private int mUrlColumnIndex;
    private int mColorColumnIndex;

    FavoriteAdapter(Context context) {
        mContext = context;
        setHasStableIds(true);
    }

    void swapCursor(Cursor cursor) {
        if (cursor == mCursor) {
            return;
        }
        if (mCursor != null) {
            mCursor.close();
        }
        mCursor = cursor;
        if (mCursor != null) {
            mIdColumnIndex = cursor.getColumnIndexOrThrow(FavoriteProvider.Columns._ID);
            mTitleColumnIndex = cursor.getColumnIndexOrThrow(FavoriteProvider.Columns.TITLE);
            mUrlColumnIndex = cursor.getColumnIndexOrThrow(FavoriteProvider.Columns.URL);
            mColorColumnIndex = cursor.getColumnIndexOrThrow(FavoriteProvider.Columns.COLOR);
        }
        notifyDataSetChanged();
    }

    @Override
    public FavoriteHolder onCreateViewHolder(ViewGroup parent, int type) {
        return new FavoriteHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_favorite, parent, false));
    }

    @Override
    public void onBindViewHolder(FavoriteHolder holder, int position) {
        if (!mCursor.moveToPosition(position)) {
            return;
        }
        long id = mCursor.getLong(mIdColumnIndex);
        String title = mCursor.getString(mTitleColumnIndex);
        String url = mCursor.getString(mUrlColumnIndex);
        int color = mCursor.getInt(mColorColumnIndex);
        holder.bind(mContext, id, title, url, color);
    }

    @Override
    public int getItemCount() {
        return mCursor != null ? mCursor.getCount() : 0;
    }

    @Override
    public long getItemId(int position) {
        return mCursor.moveToPosition(position) ? mCursor.getLong(mIdColumnIndex) : -1;
    }
}

File: app/src/main/java/org/lineageos/jelly/favorite/FavoriteAdapter.kt
/*
 * SPDX-FileCopyrightText: 2020-2024 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.jelly.favorite

import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import org.lineageos.jelly.R
import org.lineageos.jelly.model.Favorite
import org.lineageos.jelly.utils.UiUtils

class FavoriteAdapter : ListAdapter<Favorite, FavoriteAdapter.FavoriteHolder>(diffCallback) {
    var onCardClick: ((Favorite) -> Unit) = { }
    var onCardLongClick: ((Favorite) -> Unit) = { }

    override fun onCreateViewHolder(parent: ViewGroup, type: Int) = FavoriteHolder(
        LayoutInflater.from(parent.context)
            .inflate(R.layout.item_favorite, parent, false)
    )

    override fun onBindViewHolder(holder: FavoriteHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class FavoriteHolder(view: View) : RecyclerView.ViewHolder(view) {
        private val card = view.findViewById<CardView>(R.id.rowFavoriteCard)
        private val title = view.findViewById<TextView>(R.id.rowFavoriteTitle)

        fun bind(favorite: Favorite) {
            title.text = favorite.title.takeUnless {
                it.isEmpty()
            } ?: favorite.url.split("/").toTypedArray()[2]
            when (UiUtils.isColorLight(favorite.color)) {
                true -> title.setTextColor(Color.BLACK)
                false -> title.setTextColor(Color.WHITE)
            }
            card.setCardBackgroundColor(favorite.color)
            card.setOnClickListener {
                onCardClick(favorite)
            }
            card.setOnLongClickListener {
                onCardLongClick(favorite)
                true
            }
        }
    }

    companion object {
        private val diffCallback = object : DiffUtil.ItemCallback<Favorite>() {
            override fun areItemsTheSame(oldItem: Favorite, newItem: Favorite) =
                oldItem.id == newItem.id

            override fun areContentsTheSame(oldItem: Favorite, newItem: Favorite) =
                oldItem == newItem
        }
    }
}
